thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/detection/trigger/utils/DatasetTriggerInfoRepo.java

```diff
@@ -0,0 +1,165 @@
+  private static final Logger LOG = LoggerFactory.getLogger(DatasetTriggerInfoRepo.class);
+  private static DatasetTriggerInfoRepo _instance = null;
+  private static final DAORegistry DAO_REGISTRY = DAORegistry.getInstance();
+  private static Set<String> dataSourceWhitelist = new HashSet<>();
+  private static int refreshFreqInMin = 30;
+  private Map<String, Long> datasetRefreshTimeMap;
+  private ScheduledThreadPoolExecutor executorService;
+
+  private DatasetTriggerInfoRepo() {
+    this.datasetRefreshTimeMap = new ConcurrentHashMap<>();
+    this.updateFreshTimeMap(); // initial refresh
+    this.executorService = new ScheduledThreadPoolExecutor(1, r -> {
+      Thread t = Executors.defaultThreadFactory().newThread(r);
+      t.setDaemon(true);
+      return t;
+    });
+    this.executorService.scheduleAtFixedRate(
+        this::updateFreshTimeMap, refreshFreqInMin, refreshFreqInMin, TimeUnit.MINUTES);
+  }
+
+  public static void init(int refreshFreqInMin, Collection<String> dataSourceWhitelist) {
+    DatasetTriggerInfoRepo.refreshFreqInMin = refreshFreqInMin;
+    DatasetTriggerInfoRepo.dataSourceWhitelist = new HashSet<>(dataSourceWhitelist);
+  }
+
+  public static DatasetTriggerInfoRepo getInstance() {
+    if (_instance == null) {
+      synchronized (DatasetTriggerInfoRepo.class) {
+        if (_instance == null) {
+          _instance = new DatasetTriggerInfoRepo();
+        }
+      }
+    }
+    return _instance;
+  }
+
+  public boolean isDatasetActive(String dataset) {
+    return datasetRefreshTimeMap.containsKey(dataset);
+  }
+
+  public long getLastUpdateTimestamp(String dataset) {
+    return datasetRefreshTimeMap.getOrDefault(dataset, 0L);
+  }
+
+  public void setLastUpdateTimestamp(String dataset, long timestamp) {
+    datasetRefreshTimeMap.put(dataset, timestamp);
+  }
+
+  public void close() {
+    executorService.shutdown();
+    _instance = null;
+  }
+
+  private void updateFreshTimeMap() {
+    Set<String> activeDatasetSet = new HashSet<>();
+    Set<Long> visitedMetrics = new HashSet<>(); // reduce the number of DB read
+    DetectionConfigManager detectionConfigDAO = DAO_REGISTRY.getDetectionConfigManager();
+    MetricConfigManager metricsConfigDAO = DAO_REGISTRY.getMetricConfigDAO();
+    List<DetectionConfigDTO> detectionConfigs = detectionConfigDAO.findAllActive();
+    LOG.info(String.format("Found %d active detection configs", detectionConfigs.size()));
+    for (DetectionConfigDTO detectionConfig : detectionConfigs) {
+      List<String> metricUrns = DetectionConfigFormatter
+          .extractMetricUrnsFromProperties(detectionConfig.getProperties());
+      if (metricUrns.size() < 1) {
+        LOG.error("Empty metricUrn for detection {}", detectionConfig.getId());
+        continue;
+      }
+      for (String urn : metricUrns) {
+        MetricEntity me = MetricEntity.fromURN(urn);
+        if (visitedMetrics.contains(me.getId())) continue; // visited metrics
```

Comment:
We want to keep `visitedMetrics` because multiple detections might share same metric. With this set, we don't need to retrieve the metric information again.